### PR TITLE
feat(blade): added switch component

### PR DESCRIPTION
**Pull request by anuraghazra**
Changes: 

- Added Switch component, web & native
- Added tests, rn, web, ssr.
- Minor changes in Checkbox:
  - Made SelectorInput accept `hoverStyles` variant object because Checkbox & Switch have different hover styles
  - Minor changes in useCheckbox hook to add a new role: switch
  - Minor changes in SelectorLabel to add isInline prop which sets display: 'inline-block'

Screenshots & videos: 

Web: 

https://github.com/razorpay/blade/assets/35374649/daf21532-5f91-4917-b89f-d5f3924d0e49

Native: 

https://github.com/razorpay/blade/assets/35374649/63c5e7c7-49ef-46ea-9914-3dc1a1b9e42b

### CODE REVIEW

File: packages/blade/src/components/Checkbox/Checkbox.tsx [hunk 1]
```diff
@@ -213,6 +214,7 @@ const _Checkbox: React.ForwardRefRenderFunction<BladeElementRef, CheckboxProps>
         <BaseBox display="flex" flexDirection="column">
           <BaseBox display="flex" alignItems="center" flexDirection="row">
             <SelectorInput
+              hoverStyles={checkboxHoverVariants}
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Checkbox/Checkbox.tsx:217]: I see you've already thought of using hoverStlyes. Nice 😄 . Lets use that everywhere then?
```suggestion
              hoverStyles={checkboxHoverStyles}
```

File: packages/blade/src/components/Form/Selector/SelectorLabel.web.tsx [hunk 2]
```diff
@@ -4,19 +4,28 @@ import styled from 'styled-components';
...
 
-const StyledSelectorLabel = styled.label(({ theme }) => ({
-  display: 'flex',
-  marginTop: makeSpace(theme.spacing[1]),
-  marginBottom: makeSpace(theme.spacing[1]),
-}));
+const StyledSelectorLabel = styled.label<{ isInline?: boolean }>(({ theme, isInline }) => {
+  if (isInline) {
+    return {
+      display: 'inline-block',
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Form/Selector/SelectorLabel.web.tsx:10]: Who needs it inline?

File: packages/blade/src/components/Switch/AnimatedThumb.native.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isChecked]);
+
+  const thumbAnimation = useAnimatedStyle(() => {
+    return {
+      width: interpolate(
+        translateX.value,
+        [0, 0.3, 1],
+        [finalWidth, finalWidth * 1.25, finalWidth],
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/AnimatedThumb.native.tsx:63]: Add some comments explaining the multipliers here?

File: packages/blade/src/components/Switch/AnimatedThumb.web.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  }
+  100% {
+    transform: translateX(0%);
+    width: 100%;
+  }
+`;
+
+const enter = (props: { theme: DefaultTheme }) => {
+  return css`
+    ${onAnimation} ${props.theme.motion.easing.standard.effective as string} forwards;
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/AnimatedThumb.web.tsx:41]: Use castWeb instead of `as string`

File: packages/blade/src/components/Switch/AnimatedThumb.web.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  `;
+};
+const exit = (props: { theme: DefaultTheme }) => {
+  return css`
+    ${offAnimation} ${props.theme.motion.easing.standard.effective as string} forwards;
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/AnimatedThumb.web.tsx:46]: Use castWeb instead of as string

File: packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+} & SvgProps): React.ReactElement => {
+  const { theme } = useTheme();
+  const opacity = useSharedValue(isChecked ? 1 : 0);
+
+  React.useEffect(() => {
+    opacity.value = withTiming(isChecked ? 1 : 0, {
+      duration: theme.motion.duration.xquick,
+      easing: theme.motion.easing.standard.effective as EasingFunctionFactory,
+    });
+  }, [isChecked, opacity]);
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx:33]: I see you didn't add `translateX` in deps for AnimatedThumb but you've added `opacity` in deps for this. Any reason why?

File: packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+  const opacityStyle = useAnimatedStyle(() => {
+    return {
+      opacity: interpolate(opacity.value, [0, 0.2, 1], [0, 0, 1]),
+    };
+  }, []);
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx:39]: No need to add opacity as dependency here?

File: packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  return (
+    <Animated.View style={[opacityStyle]}>
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx:42]: Avoid creating a new array
```suggestion
    <Animated.View style={opacityStyle}>
```

File: packages/blade/src/components/Switch/AnimatedThumbIcon.web.tsx [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import styled from 'styled-components';
+import Svg from '~components/Icons/_Svg';
+import { makeMotionTime } from '~utils';
+
+const AnimatedThumbIcon = styled(Svg)<{ isChecked?: boolean }>(({ isChecked, theme }) => {
+  return {
+    transitionDuration: `${makeMotionTime(theme.motion.duration.xquick)}`,
+    transitionTimingFunction: theme.motion.easing.standard.effective as string,
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/AnimatedThumbIcon.web.tsx:8]: use castWeb?

File: packages/blade/src/components/Switch/Switch.tsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    value,
+    onChange,
+  });
+
+  const { theme } = useTheme();
+  const { matchedDeviceType } = useBreakpoint({ breakpoints: theme.breakpoints });
+  const previousChecked = usePrevious(state.isChecked);
+
+  return (
+    <BaseBox
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/Switch.tsx:58]: Should we make `cursor: pointer` for Switch?

File: packages/blade/src/components/Switch/Switch.tsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+      {...metaAttribute({ testID })}
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/Switch.tsx:59]: component name?

File: packages/blade/src/components/Switch/ThumbIcon.tsx [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  const finalHeight = isNumber(height) ? makeSize(height) : makeSpace(getIn(theme, height));
+  const variant = isDisabled ? 'disabled' : 'default';
+  const fillColor = getIn(theme, switchColors.thumbIcon[variant].fill);
+
+  return (
+    <AnimatedThumbIcon
+      isChecked={Boolean(isChecked)}
+      width={finalWidth}
+      height={finalHeight}
+      viewBox="0 0 11 8"
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/ThumbIcon.tsx:41]: Thats a very odd viewBox. Very different than our standard icons. Can you re-check this with design?

File: packages/blade/src/components/Checkbox/Checkbox.tsx (hunk 1 above)
Comment (anuraghazra) [reply packages/blade/src/components/Checkbox/Checkbox.tsx:217]: Looking at your previous comment, i think token makes more sense. 

hoverStyles is a bit misleading becomes that object doesn't contain any CSS styles or rules instead just blade tokens. 

Let's do <Component hoverTokens={checkboxHoverTokens} \/>

File: packages/blade/src/components/Form/Selector/types.ts [hunk 13]
```diff
@@ -1,11 +1,44 @@
...
-export type SelectorLabelProps = {
+type SelectorLabelProps = {
   children: React.ReactNode;
   /**
    * Pass only on react-native
    */
   inputProps: any; // Partial<Extract<InputProps, { onPress: any }>>;
   componentName?: string;
+  /**
+   * Changes the label to use `display: inline-block`
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Form/Selector/types.ts:12]: Removed this.

File: packages/blade/src/components/Form/Selector/types.ts [hunk 14]
```diff
@@ -1,11 +1,44 @@
...
 } & TestID;
+
+type SelectorInputProps = HoverProps & {
+  id?: string;
+  inputProps: any;
+  tabIndex?: number;
+  accessibilityLabel?: string;
+};
+
+type SelectorInputHoverVariants = {
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Form/Selector/types.ts:24]: Changed it to SelectorInputHoverTokens.

File: packages/blade/src/components/Switch/AnimatedThumb.web.tsx [hunk 15]
```diff
@@ -0,0 +1,72 @@
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Switch/AnimatedThumb.web.tsx:1]: Removed.

File: packages/blade/src/components/Switch/AnimatedThumb.web.tsx (hunk 4 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/AnimatedThumb.web.tsx:41]: Done.

File: packages/blade/src/components/Switch/AnimatedThumb.web.tsx (hunk 5 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/AnimatedThumb.web.tsx:46]: Done.

File: packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx (hunk 6 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx:33]: Yeah right I don't think its needed to add this in dependency these shared values works like refs. Otherwise this would've gotten into a infinite loop right now (setting opacity in effect and then adding opacity in the effect's deps)

File: packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx (hunk 7 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx:39]: No don't think so.

File: packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx (hunk 8 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/AnimatedThumbIcon.native.tsx:42]: Done.

File: packages/blade/src/components/Switch/Switch.tsx (hunk 10 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/Switch.tsx:58]: Might need some design inputs here, but cursor: pointer is kind of covering a bit too much of the switch's visuals

File: packages/blade/src/components/Switch/Switch.tsx (hunk 11 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/Switch.tsx:59]: Done.

File: packages/blade/src/components/Switch/ThumbIcon.tsx (hunk 12 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/ThumbIcon.tsx:41]: Yeah checked it, it's a custom SVG shape.

File: packages/blade/src/components/Switch/ThumbIcon.tsx [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+  },
+} as const;
+
+const ThumbIcon = ({
+  isChecked,
+  isDisabled,
+  size = 'medium',
+}: Pick<SwitchProps, 'isChecked' | 'isDisabled' | 'size'>): React.ReactElement => {
+  const { theme } = useTheme();
+  const { matchedDeviceType } = useBreakpoint({ breakpoints: theme.breakpoints });
```
[the remark this answers is not in the document]
Comment (anuraghazra) [reply packages/blade/src/components/Switch/ThumbIcon.tsx:27]: Yeah so the platform will contain `onDesktop` where I need `desktop`. Can do string manipulation but looks kinda hacky.

File: packages/blade/src/components/Switch/AnimatedThumb.native.tsx (hunk 3 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/AnimatedThumb.native.tsx:63]: Okay, Comes from the motion guidelines only where 

Scale thumb width by 25% basically 1.25 times

image

File: packages/blade/src/components/Switch/Switch.tsx [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+  {
+    defaultChecked,
+    isChecked,
+    isDisabled,
+    name,
+    onChange,
+    size = 'medium',
+    value,
+    accessibilityLabel,
+    id,
```
Comment (divyanshu013) [inline packages/blade/src/components/Switch/Switch.tsx:34]: We decided to remove `id` in API decision

File: packages/blade/src/components/Switch/ThumbIcon.tsx (hunk 12 above)
Comment (divyanshu013) [reply packages/blade/src/components/Switch/ThumbIcon.tsx:41]: Add a comment

File: packages/blade/src/components/Switch/__tests__/Switch.web.test.tsx [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+            isChecked={checked}
+            onChange={({ isChecked }) => setChecked(isChecked)}
+          />
+          <p data-testid="state">{checked ? 'checked' : 'unchecked'}</p>
+        </>
+      );
+    };
+    const { getByRole, getByTestId } = renderWithTheme(<Example />);
+
+    expect(getByTestId('state')).toHaveTextContent('unchecked');
```
Comment (divyanshu013) [inline packages/blade/src/components/Switch/__tests__/Switch.web.test.tsx:123]: Instead of adding a test id you can test on the text content directly, expect text content to be there

File: packages/blade/src/components/Switch/__tests__/Switch.web.test.tsx [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+  });
+
+  it('should accept testID', () => {
+    const name = 'Toggle Darkmode';
+    const { getByTestId } = renderWithTheme(
+      <Switch testID="switch-test" accessibilityLabel={name} />,
+    );
+    expect(getByTestId('switch-test')).toBeTruthy();
+  });
+});
```
Comment (divyanshu013) [inline packages/blade/src/components/Switch/__tests__/Switch.web.test.tsx:169]: Can add another assertAccessible test for web

File: packages/blade/src/tokens/global/size.ts [hunk 20]
```diff
@@ -38,6 +38,8 @@ const size = {
   36: 36,
   /** 40 px */
   40: 40,
+  /** 40 px */
```
Comment (divyanshu013) [inline packages/blade/src/tokens/global/size.ts:41]:
```suggestion
  /** 44 px */
```

**Review by divyanshu013 — COMMENTED** [review]
UX issue - when toggling with mouse, in focus state (when mouse click is in a pressed down state) the focus ring disappears and reappears after click out. Creates a slight appearance and disappearance of focus ring. Can we make it stay even in focus state?

https://github.com/razorpay/blade/assets/6682655/d4ccbffc-8ed8-46d6-8a3c-a0f34ff57533

File: packages/blade/src/components/Switch/Switch.tsx (hunk 17 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/Switch.tsx]: done.

**Comment by anuraghazra** [comment]
> UX issue - when toggling with mouse, in focus state (when mouse click is in a pressed down state) the focus ring disappears and reappears after click out. Creates a slight appearance and disappearance of focus ring. Can we make it stay even in focus state?

Ohhhh well.... it seems like a native browser behaviour, check this minmal repro: 

https://codesandbox.io/s/input-label-click-active-state-bug-6k94s7

**Comment by divyanshu013** [comment]
So no simple way to achieve this? In button the focus ring stays even in focus state

**Comment by anuraghazra** [comment]
> So no simple way to achieve this? In button the focus ring stays even in focus state

Yeah not sure really, button ka time it behaves differently it seems. 

Check this sandbox, browser literally removes the focus state from input: 

https://codesandbox.io/s/input-label-click-active-state-bug-forked-mhwgpu?file=/src/App.tsx

**Comment by anuraghazra** [comment]
OH LOL. 

onMouseDown={(e) => {
  e.preventDefault();
}}

this seems to be working 😆 let me test for edge cases.

**Comment by divyanshu013** [comment]
Non blocker though

**Comment by anuraghazra** [comment]
Works! :D 

Nice catch! 

https://github.com/razorpay/blade/assets/35374649/bba664f2-7dd7-46ae-9d45-ea017af3e781

File: packages/blade/src/components/Switch/ThumbIcon.tsx (hunk 12 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/ThumbIcon.tsx:42]: Done.

**Comment by divyanshu013** [comment]
Nice, minor: unable to toggle switch with <kbd>Enter</kbd>, <kbd>Space</kbd> works though

**Comment by anuraghazra** [comment]
> Nice, minor: unable to toggle switch with Enter, Space works though

Yes, thats expected, the Enter to toggle is optional behaviour and depends on how you've modelled your Switch, Enter is expected when you've modelled as a Button, but I've modelled it as a Checkbox input element. 

- [Modelled as Button Example](https://www.w3.org/WAI/ARIA/apg/patterns/switch/examples/switch-button/) - Enter works
- [Modelled as Input Example](https://www.w3.org/WAI/ARIA/apg/patterns/switch/examples/switch-checkbox/) - No Enter action

image

**Review by divyanshu013 — APPROVED** [review]
:shipit:

**Review by divyanshu013 — APPROVED** [review]
Looks good to me.

**Review by divyanshu013 — DISMISSED** [review]
:shipit:

File: packages/blade/src/components/Checkbox/__tests__/__snapshots__/CheckboxGroup.web.test.tsx.snap [hunk 21]
```diff
@@ -699,7 +699,7 @@ exports[`<CheckboxGroup /> should render with label 1`] = `
 }
 
 .c8:focus + div {
-  outline: 1px solid white;
+  outline: 1px solid hsla(220,30%,96%,1);
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Checkbox/__tests__/__snapshots__/CheckboxGroup.web.test.tsx.snap:702]: What is this change?

Comment (anuraghazra) [reply packages/blade/src/components/Checkbox/__tests__/__snapshots__/CheckboxGroup.web.test.tsx.snap:702]: So it's the inner outline before the focus ring, previously it was set to white which looked bad on banking dark. 

before: 
image


after:

File: packages/blade/src/components/Form/Selector/types.ts [hunk 22]
```diff
@@ -1,11 +1,49 @@
...
+  inputProps: any;
+  tabIndex?: number;
+  accessibilityLabel?: string;
+};
+
+type SelectorInputHoverTokens = {
+  default: {
+    background: {
+      checked: string;
+      unchecked: string;
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Form/Selector/types.ts:33]: Why are these strings? Shouldn't they be a type of color token?

File: packages/blade/src/components/Switch/AnimatedThumb.native.tsx [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+
+const AnimatedThumb = ({
+  isChecked,
+  isDisabled,
+  size = 'medium',
+  children,
+  isPressed,
+}: AnimatedThumbProps): React.ReactElement => {
+  const { theme } = useTheme();
+  const { matchedDeviceType } = useBreakpoint({ breakpoints: theme.breakpoints });
```
Comment (chaitanyadeorukhkar) [inline packages/blade/src/components/Switch/AnimatedThumb.native.tsx:44]:
```suggestion
  const { theme, platform } = useTheme();
```

File: packages/blade/src/components/Form/Selector/types.ts (hunk 22 above)
Comment (anuraghazra) [reply packages/blade/src/components/Form/Selector/types.ts:33]: Changed.

File: packages/blade/src/components/Switch/AnimatedThumb.native.tsx (hunk 23 above)
Comment (anuraghazra) [reply packages/blade/src/components/Switch/AnimatedThumb.native.tsx:44]: Can't because the values are `'onDesktop' | 'onMobile'` while i want `'desktop' | 'mobile'` to cleanly pick the correct value from token object.

File: packages/blade/src/components/Switch/ThumbIcon.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (chaitanyadeorukhkar) [reply packages/blade/src/components/Switch/ThumbIcon.tsx:27]: You need `desktop` because you've structured your tokens that way, make it `onDesktop` instead 😅 

Reason is that `platform` is cleanly exposed from useTheme and `onDesktop` & `onMobile` is a terminology we use on our global tokens so better to keep the same on component level tokens

**Review by chaitanyadeorukhkar — APPROVED** [review]
